Update TFM to net7.0 for source-build

lbussell opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/12000

Regression? Last working version: Not a regression.

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->
I changed all source-build conditioned occurrences of net6.0 TFMs to net7.0.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [x] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

lbussell committed:
Update TFM to net7.0 for source-build

lbussell commented:
cc @crummel

nkolev92 commented on build/common.project.props:
```diff
@@ -14,19 +14,19 @@
-    <NETCoreTargetFramework Condition="'$(DotNetBuildFromSource)' == 'true'">net6.0</NETCoreTargetFramework>
-    <NETCoreTestTargetFramework>net6.0</NETCoreTestTargetFramework>
+    <NETCoreTargetFramework Condition="'$(DotNetBuildFromSource)' == 'true'">net7.0</NETCoreTargetFramework>
+    <NETCoreTestTargetFramework>net7.0</NETCoreTestTargetFramework>
```
Please don't change the target framework in non source build scenarios. 
This can cause our builds to fail.

nkolev92 commented on build/common.targets:
```diff
@@ -6,7 +6,7 @@
     <IsDesktop>true</IsDesktop>
   </PropertyGroup>
 
-  <PropertyGroup Condition=" $(TargetFramework.StartsWith('netcoreapp')) OR $(TargetFramework.StartsWith('netstandard')) OR $(TargetFramework.StartsWith('net6')) ">
```
this would break for any projects still targeting net6. 
This check should probably do all, net5, net6 and net7.

nkolev92 commented:
@lbussell 
Can you please keep the PR template and create an issue in the home repo as per https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md#contributing-step-by-step

lbussell committed:
Don't change targetframework in non-source-build scenarios

lbussell committed:
Define IS_CORECLR for net6 target framework

lbussell commented:
> Can you please keep the PR template and create an issue in the home repo 

@nkolev92, I filed an issue, updated the PR body with the template, and addressed your review comments - is there a reason the PR validation isn't running?

nkolev92 commented:
Thanks for the updates @lbussell. 

The CI doesn't run automatically for non-NuGet members. I've applied the label and kicked off the validation.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:
```diff
@@ -12,6 +12,7 @@ public static class FrameworkConstants
         public static readonly Version MaxVersion = new Version(int.MaxValue, 0, 0, 0);
         public static readonly Version Version5 = new Version(5, 0, 0, 0);
         public static readonly Version Version6 = new Version(6, 0, 0, 0);
+        public static readonly Version Version7 = new Version(7, 0, 0, 0);
```
This is a public API change and as such it required an Unshipped updates check. 

The CI will fail because this new API is not declared.

dominoFire commented on src/NuGet.Core/NuGet.PackageManagement/NuGet.PackageManagement.csproj:
```diff
@@ -35,7 +35,7 @@
   </ItemGroup>
 
-  <ItemGroup Condition=" '$(TargetFramework)' == '$(NetStandardVersion)' or '$(TargetFramework)' == 'net6.0' ">
+  <ItemGroup Condition=" '$(TargetFramework)' == '$(NetStandardVersion)' or '$(TargetFramework)' == 'net7.0' ">
```
Just curious, why does this assembly is needed for source-build? By looking at the description, it says this assembly is for Visual Studio.

lbussell replied:
We still build NuGet.PackageManagement during source-build. We prefer to have the repo builds and the source-build be as similar as possible, so we usually only exclude things if they include pre-built binaries or have licensing issues.

As for this specific reference to System.ComponentModel.Compisition, NuGet.PackageManagement doesn't build without this reference. it looks like it's shipped in the SDK, so there's no issue referencing it here.

dominoFire reviewed (dismissed):
One question, which should not block source-build changes.

lbussell committed:
Add Version7 framework constant to public unshipped api list

dominoFire commented:
@lbussell There are some compiler errors:

```
D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\FrameworkConstants.cs(15,40): error RS0016: Symbol 'Version7' is not part of the declared API. [D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj]
D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\FrameworkConstants.cs(190,51): error RS0016: Symbol 'Net70' is not part of the declared API. [D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj]
D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\PublicAPI.Unshipped.txt(2,1): error RS0017: Symbol 'NuGet.Frameworks.FrameworkConstants.Version7 -> Version' is part of the declared API, but is either not public or could not be found [D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj]
D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\FrameworkConstants.cs(15,40): error RS0016: Symbol 'Version7' is not part of the declared API. [D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj]
D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\FrameworkConstants.cs(190,51): error RS0016: Symbol 'Net70' is not part of the declared API. [D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj]
D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\PublicAPI.Unshipped.txt(2,1): error RS0017: Symbol 'NuGet.Frameworks.FrameworkConstants.Version7 -> Version' is part of the declared API, but is either not public or could not be found [D:\a\_work\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj]
```

nkolev92 commented:
🔔 @lbussell

lbussell committed:
Merge remote-tracking branch 'upstream/dev' into net7.0-tfm

lbussell committed:
Update unshipped API format

lbussell commented:
I'm not sure if I've got the unshipped API format correct - can anyone give me some pointers?

zivkan commented:
The roslyn analyzer code fix can create the right syntax for you. You're almost certainly not using Windows and therefore Visual Studio is not an option for you, where roslyn analyzers and code fixes run by default. For reasons I don't understand, Omnisharp (or VSCode?) have them disabled by default, but in VSCode's settings I see an option "Omnisharp: Enably Roslyn Analyzers".

zivkan commented:
Also, if you build locally from the command line (cd to `src\NuGet.Core\NuGet.Frameworks`, then run `dotnet build`), the build should fail if the syntax isn't right.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:
```diff
@@ -186,6 +187,7 @@ public static class CommonFrameworks
             // .NET 5.0 and later has NetCoreApp identifier
             public static readonly NuGetFramework Net50 = new NuGetFramework(FrameworkIdentifiers.NetCoreApp, Version5);
             public static readonly NuGetFramework Net60 = new NuGetFramework(FrameworkIdentifiers.NetCoreApp, Version6);
+            public static readonly NuGetFramework Net70 = new NuGetFramework(FrameworkIdentifiers.NetCoreApp, Version7);
```
You can always just not add this change as it's not really anything that requires your change. this is merely an optimization for restore.

lbussell committed:
Update unshipped API

lbussell commented:
I was able to get the project loaded in VS and I included the updates it suggested.

lbussell commented:
Not sure what the current PR validation failure is - it doesn't look related to my changes:

https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6620811&view=logs&j=68fd25e5-7968-5c7b-bc0b-13c16b2edd08&t=2fd8befc-7898-5b33-2bea-d02ab5970aca&l=10

nkolev92 commented:
@lbussell If you rebase it should get fixed. 

Ping us here so we can trigger the CI when you've rebased.

lbussell committed:
Merge branch 'dev' into net7.0-tfm

lbussell committed:
Merge branch 'dev' into net7.0-tfm

lbussell commented:
I updated the PR again, can we run CI again?

It also looks like the last failure was a timeout - unless you think that's something my change could have caused.

lbussell commented:
@nkolev92

lbussell commented:
I also re-ran failed legs from azure devops, looks like mac didn't timeout this time.

lbussell commented:
Current test failure looks the same as this PR: https://github.com/NuGet/NuGet.Client/pull/4788
